### PR TITLE
ROC-4350 - Claimant: Paid in full - front end for all page flow

### PR DESCRIPTION
**Pull request by laurentaylor1**
### JIRA link
https://tools.hmcts.net/jira/browse/ROC-4350

### Change description
Added new feature: paid-in-full date paid page (with draft).

### Work checklist

- [ ] Unit tests added where applicable
- [x] Route tests added for new pages
- [x] New pages included in a11y tests
- [ ] Task list and task completeness checks updated
- [ ] Check and send page updated
- [x] Routes, page content and page flows of new features are toggled 
- [ ] UI changes look good on mobile
- [ ] Required Google Analytics events are being sent 

### Developer self-QA run statement

- [x] I have clicked through the running application to see if all changes I made actually work.

### Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by sabahirfan** [comment]
Please resolve conflicts.

File: src/main/features/paid-in-full/views/date-paid.njk [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+{% block content %}
+  <div class="grid-row">
+    <div class="column-two-thirds">
+      <form method="post">
+        <input type="hidden" name="known" value="true"/>
+        {{ csrfProtection(csrf) }}
+
+        <p>{{ t('Enter the exact date the money was clear in your account if you were paid by cheque or bank transfer.') }}</p>
+        {{ dateInput(form = form, name = 'date', hint = 'For example, 10 9 2018') }}
```
Comment (sabahirfan) [inline src/main/features/paid-in-full/views/date-paid.njk]: Example in hint should not be hard coded

File: src/test/features/paid-in-full/routes/checks/authorization-check.ts [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import { checkAuthorizationGuards as check } from 'test/routes/authorization-check'
+
+export function checkAuthorizationGuards (app: any, method: string, pagePath: string) {
```
Comment (sabahirfan) [inline src/test/features/paid-in-full/routes/checks/authorization-check.ts]: do we really need this wrapper?

File: src/test/features/paid-in-full/routes/date-paid.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        draftStoreServiceMock.rejectFind('Error')
+
+        await request(app)
+          .post(pagePath)
+          .set('Cookie', `${cookieName}=ABC`)
+          .send(validFormData)
+          .expect(res => expect(res).to.be.serverError.withText('Error'))
+      })
+
+      // EXPECTING 500 BUT GETS 404: expects nothing but gets draft
```
Comment (sabahirfan) [inline src/test/features/paid-in-full/routes/date-paid.ts]: please remove this comment

File: src/main/features/paid-in-full/views/date-paid.njk [hunk 4]
```diff
@@ -0,0 +1,21 @@
+{% extends "layout.njk" %}
+{% from "form.njk" import csrfProtection, errorSummary, saveAndContinueButton %}
+{% from "timelineFragment.njk" import timelineFragment %}
```
Comment (sabahirfan) [inline src/main/features/paid-in-full/views/date-paid.njk]: Please remove this unnecessary import

File: src/main/features/paid-in-full/form/models/datePaid.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
+import { IsValidLocalDate } from '@hmcts/cmc-validators'
+import { ValidateNested } from 'class-validator'
+import { LocalDate } from 'forms/models/localDate'
+import { IsNotInFuture } from 'forms/validation/validators/notInFuture'
+export class ValidationErrors {
```
Comment (sabahirfan) [inline src/main/features/paid-in-full/form/models/datePaid.ts]: Please consider if we can reuse the error messages from common place (`src/main/app/forms/validation/validationErrors.ts`)

File: src/main/features/paid-in-full/paths.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import { RoutablePath } from 'shared/router/routablePath'
+
+export const paidInFullPath = '/case/:externalId/paid-in-full'
+
+export class Paths {
+  static readonly datePaidPage = new RoutablePath(`${paidInFullPath}/date-paid`)
```
Comment (sabahirfan) [inline src/main/features/paid-in-full/paths.ts]: please add a11y tests for this as well

File: src/main/features/paid-in-full/views/date-paid.njk (hunk 1 above)
Comment (laurentaylor1) [reply src/main/features/paid-in-full/views/date-paid.njk]: Now fixed to be 2 weeks ago

File: src/test/features/paid-in-full/routes/checks/authorization-check.ts (hunk 2 above)
Comment (laurentaylor1) [reply src/test/features/paid-in-full/routes/checks/authorization-check.ts]: If I remove the wrapper and change the 'check' call to checkAuthorizationGuards, I'm given an "Import conflicts with local declaration of checkAuthorizationGuards" error

Comment (sabahirfan) [reply src/test/features/paid-in-full/routes/checks/authorization-check.ts]: I mean you don't need this file at all. Use `checkAuthorizationGuards` directly where needed

File: src/main/features/paid-in-full/views/date-paid.njk (hunk 1 above)
Comment (sabahirfan) [reply src/main/features/paid-in-full/views/date-paid.njk]: `10 9 2018` won't be two weeks ago in few months time. This need to be generated

File: src/main/features/paid-in-full/form/models/datePaid.ts (hunk 5 above)
Comment (laurentaylor1) [reply src/main/features/paid-in-full/form/models/datePaid.ts]: Changed to use common validation errors class.

File: src/main/features/paid-in-full/views/date-paid.njk (hunk 1 above)
Comment (laurentaylor1) [reply src/main/features/paid-in-full/views/date-paid.njk]: I've pushed the change to make it dynamic and calculate the example 14 days before today

Comment (sabahirfan) [reply src/main/features/paid-in-full/views/date-paid.njk]: Thanks

File: src/main/features/paid-in-full/form/models/datePaid.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  constructor (date?: LocalDate) {
+    this.date = date
+  }
+
+  static fromObject (input?: any): DatePaid {
+    if (!input) {
+      return input
+    }
+
+    const datePaid = new DatePaid(LocalDate.fromObject(input.date))
```
Comment (sabahirfan) [inline src/main/features/paid-in-full/form/models/datePaid.ts]: Don't need `const` here. simply `return new DatePaid(LocalDate.fromObject(input.date))`

File: src/main/features/paid-in-full/routes/date-paid.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+import { Paths } from 'paid-in-full/paths'
+import { ErrorHandling } from 'shared/errorHandling'
+import { Draft } from '@hmcts/draft-store-client'
+import { DraftPaidInFull } from 'paid-in-full/draft/draftPaidInFull'
+import { Form } from 'forms/form'
+import { FormValidator } from 'forms/validation/formValidator'
+import { User } from 'idam/user'
+import { DraftService } from 'services/draftService'
+import { DatePaid } from 'paid-in-full/form/models/datePaid'
+import {ClaimStoreClient} from "claims/claimStoreClient";
```
Comment (sabahirfan) [inline src/main/features/paid-in-full/routes/date-paid.ts]: please correct formatting and use single quote.

File: src/test/features/paid-in-full/routes/date-paid.ts [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+import * as draftStoreServiceMock from 'test/http-mocks/draft-store'
+import { checkNotClaimantInCaseGuard } from 'test/features/paid-in-full/routes/checks/not-claimant-in-case-check'
+import { checkAuthorizationGuards } from 'test/routes/authorization-check'
+
+const externalId = claimStoreServiceMock.sampleClaimObj.externalId
+
+const cookieName: string = config.get<string>('session.cookieName')
+const pagePath = Paths.datePaidPage.evaluateUri({ externalId: externalId })
+
+describe.only('claim - date money was received', () => {
```
Comment (sabahirfan) [inline src/test/features/paid-in-full/routes/date-paid.ts]: please remove `.only`

**Pull request merged**